api/azimuth/provider/openstack/provider.py

```
@@ -690,6 +707,79 @@ def _tenant_network(self, create_network = False):
+
+            access_list = list(self._connection.share.access.all(
+                share_id=project_share.id))
+            found_expected_access = False
+            for access in access_list:
+                if access.access_to == self._project_share_user:
+                    found_expected_access = True
+                    break
+            if not found_expected_access:
+                raise Exception("can't find the expected access rule!")
+            self._log(f"Found project share for: {self._connection.project_id}")
+
+        # no share found, create if required
+        if not project_share and create_share:
+            self._log(f"Creating project share for: {self._connection.project_id}")
+
+            # Find share type
+            default_share_type = None
+            all_types = list(self._connection.share.types.all())
+            if len(all_types) == 1:
+                default_share_type = all_types[0]
+            else:
+                for share_type in all_types:
+                    if share_type.is_default:
+                        default_share_type = share_type
+                        break
+            if not default_share_type:
+                raise Exception("Unable to find valid share type!")
+
+            # TODO(johngarbutt) need to support non-ceph types eventually
+            project_share = self._connection.share.shares.create(
+                share_proto="CephFS",
+                size=10,
+                name=self._project_share_name,
+                description="Project share auto-created by Azimuth.",
+                share_type=default_share_type.id)
+
+            # wait for share to be available before trying to grant access
+            for _ in range(10):
+                latest = self._connection.share.shares.get(project_share.id)
+                if latest.status.lower() == "available":
+                    break
+                time.sleep(0.1)
```

Does this not have an error case where the share does not become available within the allotted time?

---

It does, but that currently falls through to the next call which will fail and raise an exception.

Its lazy though... I should raise an exception when you get an error status back.